firmware/picr22-team-OWO/Core/Src/main.c

```diff
@@ -0,0 +1,879 @@
+	int16_t enc_change; // Change of the encoder, currently used for debugging purposes
+	int16_t integral; // PID stuff borrowed from kurgimopeed
+	float flat_const;
+	float int_const;
+	float der_const;
+	int16_t error;
+} Motor_Status;
+
+typedef struct Command {
+  int16_t speed[3]; // array of all of the motor speeds, CURRENTLY int16_t DUE TO TESTING WITH RAW MOTOR VALUES, MOVE BACK TO int8_t
+  uint16_t thrower_speed; // thrower motor speed
+  uint16_t servo1;
+  uint16_t servo2;
+  int16_t int_const;
+  int16_t flat_const;
+  int16_t der_const;
+  uint16_t delimiter;
+} Command;
+
+typedef struct Feedback {
+  int16_t speed[3];
+  int16_t change[3];
+  int16_t error[3];
+  int16_t integral[3];
+  uint16_t ball_detected;
+} Feedback;
+
+// Motor structs for all 3 motors
+Motor_Status motor_status[3] = {0};
+
+// Prefilled command struct
+Command command = {.speed[0] = 0,
+				   .speed[1] = 0,
+				   .speed[2] = 0,
+				   .thrower_speed = 3277,
+				   .servo1 = 4875,
+				   .servo2 = 6150,
+				   .flat_const = 1,
+				   .int_const = 0.1,
+				   .delimiter = 0};
+
+volatile uint8_t isCommandReceived = 0;
+volatile uint16_t commandless_count = 0;
+
+uint16_t clamp(uint16_t value, uint16_t min, uint16_t max) {
+	if (value > max) {return max;}
+	else if (value < min) {return min;}
+	return value;
+}
+
+
+// Copies the recieved command into the command struct
+void CDC_On_Receive(uint8_t* buffer, uint32_t* length) { // command recieve callback, copies data to command struct
+  if (*length == sizeof(Command)) {
+    memcpy(&command, buffer, sizeof(Command));
+    if (command.delimiter == 0xAAAA) {
+      isCommandReceived = 1;
+      commandless_count = 0;
+    }
+  }
+}
+
+// Initialises the encoders
+void enc_init() {
+	HAL_TIM_Encoder_Start(&htim4, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 1 ?
+	HAL_TIM_Encoder_Start(&htim1, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 2
+	HAL_TIM_Encoder_Start(&htim3, TIM_CHANNEL_1 | TIM_CHANNEL_2); // Motor 3 ?
+}
+
+// Initialises the PWMs
+void pwm_init() {
+	TIM8->CCR2 = 0;
+	TIM8->CCR1 = 0;
+	TIM8->CCR3 = 0;
+
+	TIM2->CCR1 = 4915;
+
+	TIM15->CCR1 = 4875;
+	TIM15->CCR2 = 6150;
+
+    HAL_TIM_PWM_Start(&htim8, TIM_CHANNEL_1);
+    HAL_TIMEx_PWMN_Start(&htim8, TIM_CHANNEL_2);
+    HAL_TIMEx_PWMN_Start(&htim8, TIM_CHANNEL_3);
+
+    HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+
+    HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_1);
+    HAL_TIM_PWM_Start(&htim15, TIM_CHANNEL_2);
+}
+
+// Updates the motor's speed, direction and PI constants from the command struct
+void motor_status_update() {
+	for (uint8_t i = 0; i<3; i++) {
+		if (command.speed[i] >= 0) {
+			motor_status[i].direction = 1;
+			if (command.speed[i] == 1) {command.speed[i] = 2;} // Everybody gangsta till the input speed is 1
+			motor_status[i].target_speed = command.speed[i];
+		} else {
+			motor_status[i].direction = 0;
+			if (command.speed[i] == -1) {command.speed[i] = -2;} // Everybody gangsta till the input speed is -1
```

Comment:
What's wrong with speeds 1 and -1?

Reply:
Removed it, previously it was due to extreme instability at very low speeds